src/ui/public/query_bar/directive/query_bar.js

```diff
@@ -114,7 +114,13 @@ module.directive('queryBar', function () {
 
       $scope.$watch('queryBar.indexPatterns', () => {
         this.getIndexPatterns().then(indexPatterns => {
-          this.getKuerySuggestions = getSuggestionsProvider({ $http, config, indexPatterns });
+          this.getKuerySuggestions = getSuggestionsProvider({
+            config,
+            indexPatterns: indexPatterns.map(indexPattern => ({
+              fields: indexPattern.fields.raw,
+              title: indexPattern.title,
+            })),
+          });
```

Comment:
Ideally I'd want to use index patterns directly from the saved objects API but since the fetching part is burried in the index pattern service, I just strip it down here. 
Could this cause any issues?

Reply:
I don't think it causes any technical issues atm, but having so many types of index patterns floating around makes things really confusing. It also means we have to update every usage of `getSuggestionsProvider` any time we need an additional prop from the index pattern other than `fields` and `title`.

What if we attach the saved object API response to the IndexPattern object [at creation time](https://github.com/elastic/kibana/blob/9e15904f78644d478c1507c1088cc65f2938278c/src/ui/public/index_patterns/_index_pattern.js#L210)?

Reply:
>  It also means we have to update every usage of getSuggestionsProvider any time we need an additional prop from the index pattern other than fields and title

I made a decision to make indexPattern an object:
```js
{
  fields: [...],
  title: '...'
}
```

Looking at the saved_objects API I don't know what else we'd want. But maybe we should avoid picking from saved_object and just serve it directly. Biggest caveat is that `fields` is a stringified object:

```json
{
  "id": "a5428b60-4287-11e8-94d3-0568277cdf79",
  "type": "index-pattern",
  "updated_at": "2018-04-17T21:38:20.825Z",
  "version": 2,
  "attributes": {
    "title": "apm*",
    "timeFieldName": "@timestamp",
    "fields": "..."
  }
}
```

> What if we attach the saved object API response to the IndexPattern object at creation time?
[_source](https://github.com/elastic/kibana/blob/9e15904f78644d478c1507c1088cc65f2938278c/src/ui/public/index_patterns/_index_pattern.js#L217) already contains `attributes` but if the metadata might be interesting I can add like `_saved_object` which is the unmangled `saved_object` response.

Reply:
Hmmm kinda stinks that `fields` is a string, I forgot the saved objects client does zero processing on the response. I really wish Kibana had a proper model layer. The IndexPattern class has its warts but it's much easier to work with than the raw saved object responses. The biggest problem is its dependence on angular, but it looks like decoupling it would be a colossal task. 

I'm struggling a bit to think of a good solution that doesn't make index patterns even more of a confusing mess, so I'll just throw one idea out here. Maybe we need to introduce a new lightweight IndexPattern model that can wrap around the saved object response and deserialize the `fields` attribute? We could add a method to the existing IndexPattern class that returns an instance of this lightweight model so that there's an easy and standard way to convert from old to new. Perhaps the saved object client should even return this model, or there should be a service that wraps the saved object client and returns this model. Either way, it feels like we need an additional layer of abstraction in front of the saved objects client so that our business logic isn't coupled to the saved object's serialization format.

Reply:
@Bargs I tried to to a simple transformation of the Saved Object at creation time [here](https://github.com/elastic/kibana/blob/9e15904f78644d478c1507c1088cc65f2938278c/src/ui/public/index_patterns/_index_pattern.js#L210). But it turns out discover doesn't even use that. Instead Discover fetches saved objects using the `savedObjectsClient`, and uses `courier.indexPatterns.get(id)` to get the indexPatterns: https://github.com/sqren/kibana/blob/kuery-remove-byName-dep/src/core_plugins/kibana/public/discover/controllers/discover.js#L55-L84

I felt I was getting too deep into something I really would prefer not to touch. Instead I moved the logic to the collection of static index pattern functions:  https://github.com/sqren/kibana/blob/148944eab997a4eb222394dd39e297b749833100/src/ui/public/index_patterns/static_utils/index.js